test/dynamo/test_autograd_function.py

```diff
@@ -1581,6 +1584,422 @@ def fn(x):
+            y, y_data = Foo.apply(module, input_data, x)
+            return x + y + y_data + y_data
+
+        # Call the function
+        x = torch.randn(4, 4, requires_grad=True)
+        ref = fn(input_data, x)
+
+        backend = AotEagerAndRecordGraphs()
+        opt_fn = torch.compile(fn, fullgraph=True, backend=backend)
+        res = opt_fn(input_data, x)
+        self.assertEqual(ref, res)
+
+        # Must have `view_as`
+        self.assertExpectedInline(
+            torch._dynamo.testing.normalize_gm(
+                backend.graphs[0].print_readable(print_output=False)
+            ),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_: "f32[4, 4]", L_module_b: "f32[4, 4]", L_input_data_values: "f32[4, 4]"):
+        l_x_ = L_x_
+        l_module_b = L_module_b
+        l_input_data_values = L_input_data_values
+
+        fwd_body_0 = self.fwd_body_0
+        bwd_body_0 = self.bwd_body_0
+        autograd_function_apply = torch.ops.higher_order.autograd_function_apply(fwd_body_0, bwd_body_0, l_module_b, l_input_data_values, l_x_, non_differentiable_idx = []);  fwd_body_0 = bwd_body_0 = l_module_b = l_input_data_values = None
+        getitem: "f32[4, 4]" = autograd_function_apply[0]
+        getitem_1: "f32[4, 4]" = autograd_function_apply[1];  autograd_function_apply = None
+
+        add: "f32[4, 4]" = l_x_ + getitem;  l_x_ = getitem = None
+        add_1: "f32[4, 4]" = add + getitem_1;  add = None
+        add_2: "f32[4, 4]" = add_1 + getitem_1;  add_1 = getitem_1 = None
+        return (add_2,)
+
+    class fwd_body_0(torch.nn.Module):
+        def forward(self, l_module_b: "f32[4, 4]", l_input_data_values: "f32[4, 4]", l_x_: "f32[4, 4]"):
+            _set_grad_enabled = torch._C._set_grad_enabled(False);  _set_grad_enabled = None
+
+            mul: "f32[4, 4]" = l_module_b * 5;  l_module_b = None
+            mul_1: "f32[4, 4]" = mul * l_input_data_values;  mul = None
+            output_tensor: "f32[4, 4]" = 4 + mul_1;  mul_1 = None
+
+            out: "f32[4, 4]" = output_tensor + l_x_;  output_tensor = None
+            out_1: "f32[4, 4]" = l_input_data_values + l_x_;  l_input_data_values = l_x_ = None
+
+            _set_grad_enabled_1 = torch._C._set_grad_enabled(True);  _set_grad_enabled_1 = None
+            return ((out, out_1), ())
+
+    class bwd_body_0(torch.nn.Module):
+        def forward(self, grad_output: "f32[4, 4]", output_data: "f32[4, 4]"):
+            _set_grad_enabled = torch._C._set_grad_enabled(False);  _set_grad_enabled = None
+
+            mul: "f32[4, 4]" = grad_output * 4;  mul = None
+            mul_1: "f32[4, 4]" = grad_output * 2;  grad_output = None
+
+            _set_grad_enabled_1 = torch._C._set_grad_enabled(True);  _set_grad_enabled_1 = None
+            return (None, None, mul_1)
+""",
+        )
+
+    # Requires fixing issue - https://github.com/pytorch/pytorch/issues/168395
+    @unittest.expectedFailure
+    def test_nn_module_pytrees_as_io(self):
```

Comment:
erm, for what its worth, dataclasses aren't pytrees by default. one needs to register as pytrees. So the name of this test and the previous one are a  bit confusing

Reply:
Changed